keras_cv/models/densenet.py

```diff
@@ -0,0 +1,296 @@
+
+Reference:
+  - [Densely Connected Convolutional Networks](https://arxiv.org/abs/1608.06993)
+"""
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import backend
+from tensorflow.keras import layers
+
+BN_AXIS = 3
+
+
+def dense_block(x, blocks, name):
+    """A dense block.
+
+    Args:
+      x: input tensor.
+      blocks: integer, the number of building blocks.
+      name: string, block label.
+
+    Returns:
+      Output tensor for the block.
+    """
+    for i in range(blocks):
+        x = conv_block(x, 32, name=name + "_block" + str(i + 1))
+    return x
+
+
+def transition_block(x, reduction, name):
+    """A transition block.
+
+    Args:
+      x: input tensor.
+      reduction: float, compression rate at transition layers.
+      name: string, block label.
+
+    Returns:
+      output tensor for the block.
+    """
+    x = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_bn")(x)
+    x = layers.Activation("relu", name=name + "_relu")(x)
+    x = layers.Conv2D(
+        int(backend.int_shape(x)[BN_AXIS] * reduction),
+        1,
+        use_bias=False,
+        name=name + "_conv",
+    )(x)
+    x = layers.AveragePooling2D(2, strides=2, name=name + "_pool")(x)
+    return x
+
+
+def conv_block(x, growth_rate, name):
+    """A building block for a dense block.
+
+    Args:
+      x: input tensor.
+      growth_rate: float, growth rate at dense layers.
+      name: string, block label.
+
+    Returns:
+      Output tensor for the block.
+    """
+    x1 = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_0_bn")(
+        x
+    )
+    x1 = layers.Activation("relu", name=name + "_0_relu")(x1)
+    x1 = layers.Conv2D(4 * growth_rate, 1, use_bias=False, name=name + "_1_conv")(x1)
+    x1 = layers.BatchNormalization(axis=BN_AXIS, epsilon=1.001e-5, name=name + "_1_bn")(
+        x1
+    )
+    x1 = layers.Activation("relu", name=name + "_1_relu")(x1)
+    x1 = layers.Conv2D(
+        growth_rate, 3, padding="same", use_bias=False, name=name + "_2_conv"
+    )(x1)
+    x = layers.Concatenate(axis=BN_AXIS, name=name + "_concat")([x, x1])
+    return x
+
+
+def _get_name(blocks):
+    if blocks == [6, 12, 24, 16]:
+        return "densenet121"
+    if blocks == [6, 12, 32, 32]:
+        return "densenet169"
+    if blocks == [6, 12, 48, 32]:
+        return "densenet201"
+    return "densenet"
+
+
+def _apply_classifier(x, classes, classifier_activation):
+    x = layers.GlobalAveragePooling2D(name="avg_pool")(x)
+    x = layers.Dense(classes, activation=classifier_activation, name="predictions")(x)
+    return x
+
+
+def _apply_pooling_layer(x, pooling):
+    print(x.shape)
+    if pooling == "avg":
+        return layers.GlobalAveragePooling2D(name="avg_pool")(x)
+    if pooling == "max":
+        return layers.GlobalMaxPooling2D(name="max_pool")(x)
+    return x
+
+
+def DenseNet(
+    blocks,
+    include_rescaling,
+    include_top,
```

Comment:
Only for discussion: is classification head really needed, what would be the biggest use case?

Reply:
I think classification head serves the purpose of using a keras_cv.model in a fully standalone way.

i.e., without this every user of keras_cv.models would have to wrap a model in another sequential model or Functional model to perform classification.

maybe this is fine, but I think allowing them for standalone use is quite nice - particularly in tutorials.  A tutorial on CutMix should only highlight CutMix, and ideally the classifier should be a one liner.